targets/targets.json

```diff
@@ -4217,5 +4217,47 @@
         "inherits": ["FVP_MPS2"],
         "core": "Cortex-M7",
         "macros": ["CMSDK_CM7"]
+    },
+    "NUMAKER_PFM_M2351": {
+        "core": "Cortex-M23-NS",
+        "default_toolchain": "GCC_ARM",
+        "extra_labels": ["NUVOTON", "M2351", "M2351KIAAEES", "FLASH_CMSIS_ALGO"],
+        "OUTPUT_EXT": "hex",
+        "macros": ["MBED_FAULT_HANDLER_DISABLED", "MBED_TZ_DEFAULT_ACCESS=1", "LPTICKER_DELAY_TICKS=3"],
+        "is_disk_virtual": true,
+        "supported_toolchains": ["GCC_ARM", "IAR", "ARMC6"],
+        "config": {
+            "gpio-irq-debounce-enable": {
+                "help": "Enable GPIO IRQ debounce",
+                "value": 0
+            },
+            "gpio-irq-debounce-enable-list": {
+                "help": "Comma separated pin list to enable GPIO IRQ debounce",
+                "value": "NC"
+            },
+            "gpio-irq-debounce-clock-source": {
+                "help": "Select GPIO IRQ debounce clock source: GPIO_DBCTL_DBCLKSRC_HCLK or GPIO_DBCTL_DBCLKSRC_LIRC",
+                "value": "GPIO_DBCTL_DBCLKSRC_LIRC"
+            },
+            "gpio-irq-debounce-sample-rate": {
+                "help": "Select GPIO IRQ debounce sample rate: GPIO_DBCTL_DBCLKSEL_1, GPIO_DBCTL_DBCLKSEL_2, GPIO_DBCTL_DBCLKSEL_4, ..., or GPIO_DBCTL_DBCLKSEL_32768",
+                "value": "GPIO_DBCTL_DBCLKSEL_16"
+            }
+        },
+        "inherits": ["Target"],
+        "device_has": ["USTICKER", "LPTICKER", "RTC", "ANALOGIN", "I2C", "I2CSLAVE", "I2C_ASYNCH", "INTERRUPTIN", "PORTIN", "PORTINOUT", "PORTOUT", "PWMOUT", "SERIAL", "SERIAL_ASYNCH", "SERIAL_FC", "STDIO_MESSAGES", "SLEEP", "SPI", "SPISLAVE", "SPI_ASYNCH", "TRNG", "FLASH"],
+        "detect_code": ["1305"],
+        "release_versions": ["5"],
+        "device_name": "M2351KIAAEES",
+        "bootloader_supported": true
+    },
+    "NUMAKER_PFM_M2351_S": {
+        "core": "Cortex-M23",
+        "inherits": ["NUMAKER_PFM_M2351"],
+        "device_has_remove": ["TRNG"]
+    },
+    "NUMAKER_PFM_M2351_NS": {
```

Comment:
Do we need `NUMAKER_PFM_M2351_NS`? Can we use `NUMAKER_PFM_M2351` only for non-secure?

Reply:
Could we not have the `_S` variant either, and rely on the `target.core` override.

Reply:
`_S`  variant is required, else in secure examples we will have to suggest code change in targets.json which is not a good idea.

`_NS` is actually the duplicate, hence I suggested to remove that.

Reply:
A few things:
 * `target.core` is an allowed override in `mbed_app.json`
 * `target.device_has` is also override-able in `mbed_app.json`
 * The difference between `_S` and `_NS` is `target.core` and `target.device_has`
 * There are core-specific scan rules available

This implies that the base target can be converted into the 2 variants entirely with `mbed_app.json`.

I recommend that we default to Non-Secure, and allow users to override or create a custom target for compiler secure partitions.

Reply:
**NUMAKER_PFM_M2351** defaults to non-secure target.
**NUMAKER_PFM_M2351_S** means secure target.
**NUMAKER_PFM_M2351_NS** means non-secure target and is duplicate of **NUMAKER_PFM_M2351**.

The presence of **NUMAKER_PFM_M2351_S**/**NUMAKER_PFM_M2351_NS** is to avoid ambiguous with just **NUMAKER_PFM_M2351**. Besides, it may not so straightforward for user to create these variants in `mbed_app.json`.

Reply:
@ccli8 This creates confusion: As a user, I have to know what variant I'm using instead of "pick my board and go". We have the same complaint with the NRF51 based targets and their variants. I would find it especially confusing if one of the variants is a duplicate!

You imply that it's not simple to add these changes to `mbed_app.json`, yet the targets differ by core and a _single_ entry in `device_has`. Have you tried it? What makes you say that it would be difficult?

Reply:
@ccli8 - Also we do not want user to think Mbed OS can be secure, _S variant is only to help in development of secure side, it should not be the final secure application for a target.  

_S in Mbed OS code will give false impression to users. We expect target owners to place a secure binary / cmse library and any required header file in target specific folder for Nuvoton M2351 it should be `TARGET_NUVOTON\TARGET_M2351\TARGET_NUMAKER_PFM_M2351`

Reply:
I'm going to continue to say that the `_S` variant is not useful, as the following `mbed_app.json` does the same thing:
```json
{
    "target_overrides": {
        "NUMAKER_PFM_M2351": {
            "target.core": "Cortex-M23",
            "target.device_have_remove": ["TRNG"]
        }
    }
}
```
anyone that needs to switch often can add the target into `custom_targets.json` or create multiple application configurations.